url/url.pyx

```diff
@@ -0,0 +1,265 @@
+            return self.ptr.fragment()
+        def __set__(self, s):
+            self.ptr.setFragment(s)
+
+    property userinfo:
+        def __get__(self):
+            return self.ptr.userinfo()
+        def __set__(self, s):
+            self.ptr.setUserinfo(s)
+
+    def copy(self):
+        '''Return a new instance of an identical URL.'''
+        new = URL('')
+        new.ptr.setScheme(self.ptr.scheme())
+        new.ptr.setHost(self.ptr.host())
+        new.ptr.setPort(self.ptr.port())
+        new.ptr.setPath(self.ptr.path())
+        new.ptr.setParams(self.ptr.params())
+        new.ptr.setQuery(self.ptr.query())
+        new.ptr.setFragment(self.ptr.fragment())
+        new.ptr.setUserinfo(self.ptr.userinfo())
+        return new
+
+    def equiv(self, other, encoding='utf-8'):
+        '''Return true if this url is equivalent to another'''
+        if isinstance(other, basestring):
+            _other = self.parse(other, encoding)
+        else:
+            _other = self.parse(other.utf8)
+
+        _self = self.parse(self.utf8, 'utf-8')
+        _self.canonical().defrag().abspath().escape().punycode().remove_default_port()
+        _other.canonical().defrag().abspath().escape().punycode().remove_default_port()
+
+        return (
+            _self.scheme    == _other.scheme    and
+            _self.host      == _other.host      and
+            _self.path      == _other.path      and
+            _self.params    == _other.params    and
+            _self.query     == _other.query     and
+            _self.port      == _other.port
+        )
+
+    def __richcmp__(self, other, op):
+        '''Return true if this url is /exactly/ equal to another'''
+        if op == 2:  # ==
+            if isinstance(other, basestring):
+                return self.__eq__(self.parse(other, 'utf-8'))
+            return (
+                self.scheme    == other.scheme    and
+                self.host      == other.host      and
+                self.path      == other.path      and
+                self.port      == other.port      and
+                self.params    == other.params    and
+                self.query     == other.query     and
+                self.fragment  == other.fragment  and
+                self.userinfo  == other.userinfo)
+        elif op == 3:  # !=
+            return not (self == other)
+        else:
+            raise NotImplementedError(
+                '%s does not support this operation.' % type(self).__name__)
+
+    def __str__(self):
+        return self.utf8
+
+    def __repr__(self):
+        return '<url.URL object "%s" >' % self.utf8
+
+    def canonical(self):
+        self.ptr.sort_query()
+        return self
+
+    def defrag(self):
+        '''Remove the fragment from this url'''
+        self.ptr.defrag()
+        return self
+
+    def deparam(self, params):
+        '''Strip any of the provided parameters out of the url'''
+        lowered = unordered_set[string](p.lower() for p in params)
```

Comment:
That's not ringing a bell. I know we talked about casing in the context of nameprep casemapping stuff, but my memory was that we were content with continuing the use of case-insensitive deparams.